llm_on_ray/inference/predictor_deployment.py

```diff
@@ -93,145 +104,253 @@ async def consume_streamer_async(self, streamer: TextIteratorStreamer):
+                        num_generated_tokens=1,
+                        preprocessing_time=0,
+                    )
+                    yield model_response
+        else:  # use transformers predictor
             streamer = self.predictor.get_streamer()
             self.loop.run_in_executor(
                 None,
-                functools.partial(self.predictor.streaming_generate, prompts, streamer, **config),
-            )
-            return StreamingResponse(
-                self.consume_streamer_async(streamer), status_code=200, media_type="text/plain"
+                functools.partial(self.predictor.streaming_generate, prompt, streamer, **config),
             )
 
-    async def openai_call(self, prompt, config, streaming_response=True):
-        prompts = []
-        images = []
-        if isinstance(prompt, list):
-            prompt_format = get_prompt_format(prompt)
+            if not self.use_openai:
+                yield StreamingResponse(
+                    self.consume_streamer_async(streamer), status_code=200, media_type="text/plain"
+                )
+            else:
+                async for output in self.consume_streamer_async(streamer):
+                    model_response = ModelResponse(
+                        generated_text=output,
+                        num_input_tokens=self.predictor.input_length,
+                        num_input_tokens_batch=self.predictor.input_length,
+                        num_generated_tokens=1,
+                        preprocessing_time=0,
+                    )
+                    yield model_response
+
+    # Handle non-streaming, support single and multiple prompts
+    async def handle_non_streaming(self, prompts, config) -> Union[JSONResponse, str]:
+        # Use vllm for continuous batching
+        if self.use_vllm:
+            return await self.predictor.generate_async(prompts, **config)
+        else:
+            # static batching
+            if isinstance(prompts, list):
+                return await self.handle_static_batch(prompts, **config)
+            # dynamic batching
+            return await self.handle_dynamic_batch((prompts, config))
+
+        return JSONResponse(status_code=400, content="Error when handling non-streaming request.")
+
+    # TODO: get max_batch_size from the serve config
+    @serve.batch(max_batch_size=4)
+    async def handle_dynamic_batch(self, requests):
+        logger.info(f"Handling dynamic batch (size={len(requests)}) ...")
+
+        # batch prompts that configs are the same and also save request index
+        batched_prompts: Dict[str, Tuple[Union[str, List[str]], List[int]]] = {}
+        for i, request in enumerate(requests):
+            prompt = request[0]
+            config = request[1]
+            # sort the config by key and convert to str to ensure the key is unique
+            key = str(dict(sorted(config.items())))
+            batched_prompts[key] = batched_prompts.get(str(config), ([], []))
```

Comment:
This should be str(dict(sorted(config.items()))), now if config is different from sorted(config.items()), batched_prompts[key] will have only one prompt.